api/post.go

```diff
@@ -183,6 +185,86 @@ func updatePost(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func pinPost(c *Context, w http.ResponseWriter, r *http.Request) {
+	params := mux.Vars(r)
+
+	channelId := params["channel_id"]
+	if len(channelId) != 26 {
+		c.SetInvalidParam("pinPost", "channelId")
+		return
+	}
+
+	if !HasPermissionToChannelContext(c, channelId, model.PERMISSION_READ_CHANNEL) {
+		return
+	}
+
+	postId := params["post_id"]
+	if len(postId) != 26 {
+		c.SetInvalidParam("pinPost", "postId")
+		return
+	}
+
+	pchan := app.Srv.Store.Post().Get(postId)
+
+	if result := <-pchan; result.Err != nil {
+		c.Err = result.Err
+		return
+	} else if post := result.Data.(*model.PostList).Posts[postId]; post.ChannelId != channelId {
+		c.Err = model.NewLocAppError("pinPost", "api.post.pin_post.mismatched_channel_id.app_error",
+			nil, "channelId="+channelId+", post.ChannelId="+post.ChannelId+", postId="+postId)
+		c.Err.StatusCode = http.StatusBadRequest
+		return
+	}
+
+	if result := <-app.Srv.Store.Post().Pin(postId); result.Err != nil {
+		c.Err = result.Err
+		return
+	} else {
+		post := result.Data.(*model.PostList).Posts[postId]
+		w.Write([]byte(post.ToJson()))
+	}
+}
+
+func unpinPost(c *Context, w http.ResponseWriter, r *http.Request) {
+	params := mux.Vars(r)
+
+	channelId := params["channel_id"]
+	if len(channelId) != 26 {
+		c.SetInvalidParam("unpinPost", "channelId")
+		return
+	}
+
+	if !HasPermissionToChannelContext(c, channelId, model.PERMISSION_READ_CHANNEL) {
+		return
+	}
+
+	postId := params["post_id"]
+	if len(postId) != 26 {
+		c.SetInvalidParam("unpinPost", "postId")
+		return
+	}
+
+	pchan := app.Srv.Store.Post().Get(postId)
+
+	if result := <-pchan; result.Err != nil {
+		c.Err = result.Err
+		return
+	} else if post := result.Data.(*model.PostList).Posts[postId]; post.ChannelId != channelId {
+		c.Err = model.NewLocAppError("unpinPost", "api.post.unpin_post.mismatched_channel_id.app_error",
+			nil, "channelId="+channelId+", post.ChannelId="+post.ChannelId+", postId="+postId)
+		c.Err.StatusCode = http.StatusBadRequest
+		return
+	}
+
+	if result := <-app.Srv.Store.Post().Unpin(postId); result.Err != nil {
+		c.Err = result.Err
+		return
+	} else {
+		post := result.Data.(*model.PostList).Posts[postId]
```

Comment:
Hi @Zhouzi, what about instead of having specific SQL functions and queries that update the post, you get the post first, modify it, then update it with the existing functionality? That way you will have the post to return and there's less code overall since you don't need to add new SQL functions